internal/abci/cometbft/peers.go

```
@@ -1,10 +1,183 @@
+// Static peers are those that are added in the genesis and are trusted by default. These
+// include the initial validators, persistent peers, and seed nodes.
+// Whitelist peers are the list of allowed sentry nodes that a node can connect to.
+// Whitelist peers can be updated dynamically using kwil-admin commands and are persisted
+// to disk.
+// The Peers gossiped using PEX are not automatically trusted, and to be added manually if needed.
+// Any new validators are automatically added to the whitelist peers and demoted validators are removed from the whitelist.
+type Peers struct {
+	// whitelistPeers is a map of node IDs.
+	whitelistPeers map[string]bool
+	peerMtx        sync.RWMutex // protects peers
+
+	removePeerFn func(peerID string) error
+
+	db sql.TxMaker
+}
+
+func initTables(ctx context.Context, tx sql.DB) error {
+	_, err := tx.Execute(ctx, initPeersTable)
+	return err
+}
+
+// NewPeers creates a new Peers object.
+func P2PInit(ctx context.Context, db sql.TxMaker, whitelistPeers []string) (*Peers, error) {
+	upgradeFns := map[int64]versioning.UpgradeFunc{
+		0: initTables,
+	}
+
+	if err := versioning.Upgrade(ctx, db, p2pSchemaName, upgradeFns, p2pStoreVersion); err != nil {
+		return nil, err
+	}
+
+	p := &Peers{
+		whitelistPeers: make(map[string]bool),
+		db:             db,
+	}
+
+	// no need to persist them and these are added only once at startup
+	for _, peer := range whitelistPeers {
+		p.whitelistPeers[peer] = true
+	}
+
+	// Load peers from disk
+	if err := p.loadPeers(ctx); err != nil {
+		return nil, err
+	}
+
+	// load the whitelist peers from the config and persist if not already present
+	for _, peer := range whitelistPeers {
+		p.AddPeer(ctx, peer)
+	}
+
+	return p, nil
+}
+
+// AddPeer adds a peer to the Peers object.
+func (p *Peers) AddPeer(ctx context.Context, peer string) error {
+	p.peerMtx.Lock()
+	defer p.peerMtx.Unlock()
+
+	_, ok := p.whitelistPeers[peer]
+	if ok {
+		return nil
```

Noted in the ON CONFLICT clause, but I think we should return an error here